File: modulo2/programasGabarito/falsa_posicao_modificado/cpp/main.cpp
#include <iostream>
#include <iomanip>
#include <cmath>
#include <string>

int main() {
    using std::cout;
    using std::cin;
    using std::endl;

    // Variáveis (float para reproduzir o "real" simples do Fortran)
    float xi = 0.0f, xs = 0.0f;     // intervalo inferior e superior
    float xi1 = 0.0f, xs1 = 0.0f;   // intervalos na iteração anterior
    float xm = 0.0f, xm1 = 0.0f;    // ponto médio atual e anterior
    float f1 = 0.0f, f2 = 0.0f, f3 = 0.0f; // valores da função
    float produto = 0.0f;
    float erro = 0.0f;
    float tol = 0.0f;
    float delta = 0.0f;             // kick usado para desprender extremidades
    int iter = 0;
    int opcao = 0;
    int contador1 = 0, contador2 = 0;

    cout << "================================================================================" << endl;
    cout << "*                                                                              *" << endl;
    cout << "*                   L2C : SOLUÇÕES EM COMPUTAÇÃO CIENTÍFICA                    *" << endl;
    cout << "*                                                                              *" << endl;
    cout << "================================================================================" << endl;
    cout << "* Programa: RAÍZES DE EQUAÇÕES – FALSA POSIÇÃO MODIFICADO                      *" << endl;
    cout << "* Capítulo: 02 | Exemplo: 02                                                   *" << endl;
    cout << "*                                                                              *" << endl;
    cout << "* Repositório: Manual de Cálculo Numérico – Códigos de Apoio (L2C)             *" << endl;
    cout << "* Curso: Do Cálculo à Simulação Computacional                                  *" << endl;
    cout << "*                                                                              *" << endl;
    cout << "*        https://github.com/l2c-dev/manual-calculo-numerico.git                *" << endl;
    cout << "================================================================================" << endl;
    cout << "" << endl;
    cout << "================================================================================" << endl;
    cout << "*                                                                              *" << endl;
    cout << "* Esse programa resolve a equação algébrica x**10 - 1.0.                       *" << endl;
    cout << "* Nesse caso, a solução exata é dada por x = 1.                                *" << endl;
    cout << "* A ideia é mostrar como o método da falsa posição pode ficar preso em uma das *" << endl;
    cout << "* extremidades dependendo da cara da função, cujo zero estamos buscando.       *" << endl;
    cout << "* Teste usar o seguinte intervalo de busca inicial=[0,1.8].                    *" << endl;
    cout << "* Compare o desempenho do método da bissecção com o método da falsa posição.   *" << endl;
    cout << "* Em seguida, teste o método da falsa posição modificado com um kick = 0.06    *" << endl;
    cout << "*                                                                              *" << endl;
    cout << "* Quais são suas conclusões?                                                   *" << endl;
    cout << "================================================================================" << endl;
    cout << "" << endl;
    cout << "================================================================================" << endl;
    cout << "*                           SELEÇÃO DE MÉTODO                                  *" << endl;
    cout << "================================================================================" << endl;
    cout << "* Escolha agora o método a ser utilizado:                                      *" << endl;
    cout << "*                                                                              *" << endl;
    cout << "*   [1] Método da bissecção                                                    *" << endl;
    cout << "*   [2] Método da falsa posição                                                *" << endl;
    cout << "*   [3] Método da falsa posição modificado                                     *" << endl;
    cout << "*                                                                              *" << endl;
    cout << "================================================================================" << endl;
    cout << "Digite o número do método desejado: " << endl;
    cin  >> opcao;
    cout << "" << endl;
    cout << "================================================================================" << endl;
    cout << "*                 INTERVALO INICIAL DE BUSCA                                   *" << endl;
    cout << "================================================================================" << endl;
    cout << "Digite agora o extremo inferior do intervalo inicial de busca (xi):" << endl;
    cin  >> xi;
    cout << "Digite agora o extremo superior do intervalo inicial de busca (xs):" << endl;
    cin  >> xs;
    cout << "*                                                                              *" << endl;
    if (opcao == 3) {
        cout << "*Digite a intensidade do kick utilizado para desprender extremidades:        *" << endl;
        cin  >> delta;
    }
    cout << "================================================================================" << endl;
    cout << "" << endl;

    switch (opcao) {
        case 1:
            cout << "O método escolhido é o método da bisecção" << endl;
            break;
        case 2:
            cout << "O método escolhido é o método da falsa posição" << endl;
            break;
        case 3:
            cout << "O método escolhido é o método da falsa posição modificado" << endl;
            break;
        default:
            cout << "Opção inválida." << endl;
            break;
    }

    cout << "" << endl;
    cout << "================================================================================" << endl;
    cout << "*         IMPRIMINDO VARIÁVEIS DE MONITORAMENTO: iter, xi, xs, xr, erro        *" << endl;
    cout << "================================================================================" << endl;

    erro = std::fabs(xi - xs);
    tol  = 1.0e-06f;
    iter = 0;
    contador1 = 1;
    contador2 = 1;

    while (erro > tol) {
        // Formato: (I4,5X,F13.9,5X,F13.9,5X,F13.9,5X,F13.9)
        cout << std::setw(4) << iter
             << std::string(5, ' ')
             << std::setw(13) << std::fixed << std::setprecision(9) << xi
             << std::string(5, ' ')
             << std::setw(13) << std::fixed << std::setprecision(9) << xs
             << std::string(5, ' ')
             << std::setw(13) << std::fixed << std::setprecision(9) << xm
             << std::string(5, ' ')
             << std::setw(13) << std::fixed << std::setprecision(9) << erro
             << endl;

        xm1 = xm;

        // f(x) = x**10 - 1.0
        f1 = std::pow(xi, 10.0f) - 1.0f; // f(xi)
        f2 = std::pow(xs, 10.0f) - 1.0f; // f(xs)

        // Produto f(xi)*f(xs)
        produto = f1 * f2;

        // Checando mudança de sinal
        if (produto < 0.0f) {
            if (opcao == 1) {
                // Bissecção
                xm = (xi + xs) * 0.5f;
            }

            if (opcao == 2) {
                // Falsa posição
                xm = xs - (f2 * (xi - xs) / (f1 - f2));
            }

            if (opcao == 3) {
                // Falsa posição modificada
                xi1 = xi;
                xs1 = xs;

                // Verificando extremidade presa
                if (contador1 == 2) {
                    xi += delta;
                    contador1 = 1;
                }
                if (contador2 == 2) {
                    xs -= delta;
                    contador2 = 1;
                }

                // Falsa posição
                xm = xs - (f2 * (xi - xs) / (f1 - f2));
            }

            // f(xm)
            f3 = std::pow(xm, 10.0f) - 1.0f;

            // Novo produto para atualizar o intervalo
            produto = f1 * f3;

            if (produto < 0.0f) {
                xs = xm;
            } else {
                xi = xm;
            }
        }

        if (opcao == 3) {
            // Atualizando contadores de repetição das extremidades
            if (xi1 == xi) {
                contador1 = contador1 + 1;
            }
            if (xs1 == xs) {
                contador2 = contador2 + 1;
            }
        }

        erro = std::fabs(xm1 - xm);
        iter = iter + 1;
    }

    return 0;
}

File: modulo2/programasGabarito/falsa_posicao_modificado/fortran/main.f90
program falsa_posicao_modificado

implicit none

real xi, xs ! intervalo inferior e superior
real xi1, xs1 ! intervalos inferior e superior na iteração anterior
real xm, xm1 ! ponto médio atual e ponto médio antes da atualização
real f1 , f2, f3 ! função que desejamos minimizar nos pontos mínimo, máximo e médio
real produto ! produto f(xi)*f(xs)
real erro ! erro relativo
real tol ! tolerancia desejada
real delta ! kick usado para desprender uma das extremidades;
integer iter ! número da iteração
integer opcao ! opcao de metodo desejado
integer contador1 ! contador que vai contar o número de vezes que xi se repete
integer contador2 ! contador que vai contar o número de vezes que xs se repete

print *, '================================================================================'
print *, '*                                                                              *'
print *, '*                   L2C : SOLUÇÕES EM COMPUTAÇÃO CIENTÍFICA                    *'
print *, '*                                                                              *'
print *, '================================================================================'
print *, '* Programa: RAÍZES DE EQUAÇÕES – FALSA POSIÇÃO MODIFICADO                      *'
print *, '* Capítulo: 02 | Exemplo: 02                                                   *'
print *, '*                                                                              *'
print *, '* Repositório: Manual de Cálculo Numérico – Códigos de Apoio (L2C)             *'
print *, '* Curso: Do Cálculo à Simulação Computacional                                  *'
print *, '* Autor: Prof. Rafael Gabler Gontijo, PhD                                      *'
print *, '*                                                                              *'
print *, '*        https://github.com/l2c-dev/manual-calculo-numerico.git                *'
print *, '================================================================================'
print *, ''
print *, '================================================================================'
print *, '*                                                                              *'
print *, '* Esse programa resolve a equação algébrica x**10 - 1.0.                       *'
print *, '* Nesse caso, a solução exata é dada por x = 1.                                *'
print *, '* A ideia é mostrar como o método da falsa posição pode ficar preso em uma das *' 
print *, '* extremidades dependendo da cara da função, cujo zero estamos buscando.       *'
print *, '* Teste usar o seguinte intervalo de busca inicial=[0,1.8].                    *' 
print *, '* Compare o desempenho do método da bissecção com o método da falsa posição.   *'
print *, '* Em seguida, teste o método da falsa posição modificado com um kick = 0.06    *'
print *, '*                                                                              *'
print *, '* Quais são suas conclusões?                                                   *'
print *, '================================================================================'
print *, ''
print *, '================================================================================'
print *, '*                           SELEÇÃO DE MÉTODO                                  *'
print *, '================================================================================'
print *, '* Escolha agora o método a ser utilizado:                                      *'
print *, '*                                                                              *'
print *, '*   [1] Método da bissecção                                                    *'
print *, '*   [2] Método da falsa posição                                                *'
print *, '*   [3] Método da falsa posição modificado                                     *'
print *, '*                                                                              *'
print *, '================================================================================'
print *, 'Digite o número do método desejado: '
read*,opcao
print *, ''
print *, '================================================================================'
print *, '*                 INTERVALO INICIAL DE BUSCA                                   *'
print *, '================================================================================'
print *, 'Digite agora o extremo inferior do intervalo inicial de busca (xi):'; read *, xi
print *, 'Digite agora o extremo superior do intervalo inicial de busca (xs):'; read *, xs
print *, '*                                                                              *'
if(opcao.eq.3) then
write(*,*) '*Digite a intensidade do kick utilizado para desprender extremidades:        *'
read*,delta
end if
print *, '================================================================================'
print *, ''
select case (opcao)
case (1)
   write(*,'(A)') 'O método escolhido é o método da bisecção'
case (2)
   write(*,'(A)') 'O método escolhido é o método da falsa posição'
case (3)
   write(*,'(A)') 'O método escolhido é o método da falsa posição modificado'
case default
   write(*,'(A)') 'Opção inválida.'
end select
print *, ''
print *, '================================================================================'
print *, '*         IMPRIMINDO VARIÁVEIS DE MONITORAMENTO: iter, xi, xs, xr, erro        *'
print *, '================================================================================'


erro = abs(xi-xs)
tol = 1.0e-06
iter = 0
contador1 = 1
contador2 = 1

do while (erro.gt.tol) 

! Imprimindo na tela as variáveis de monitoramento abaixo

write(*,'(I4,5X,F13.9,5X,F13.9,5X,F13.9,5X,F13.9)') iter, xi, xs, xm, erro
xm1=xm
! Avaliando inicialmente a função f em cada ponto do intervalo

f1 = xi**(10.0) - 1.0 ! f(xi)
f2 = xs**(10.0) - 1.0 ! f(xs)

! Calculando o produto entre a função nos extermos do intervalo

produto = f1*f2

! Checando se houve mudança de sinal nesse intervalo
if (produto.lt.0.0) then

if(opcao.eq.1) then 
! Utiliza-se a fórmula para estimativa da raiz via método da bisecção
xm = (xi+xs)/2
end if

if(opcao.eq.2) then
! Utiliza-se a fórmula para estimativa da raiz via falsa-posicao
xm = xs - (f2*(xi-xs)/(f1-f2))
end if

if(opcao.eq.3) then
xi1 = xi
xs1 = xs
! Verificando se existe uma extremidade presa e agindo caso haja
if(contador1.eq.2) then
xi=xi+delta
contador1=1
end if
if(contador2.eq.2) then
xs=xs-delta
contador2=1
end if
! Utilizando a fórmula para estimativa da raiz via falsa-posicao
xm = xs - (f2*(xi-xs)/(f1-f2))
end if
! Estimando a função no ponto corrigido entre o intervalo
f3 = xm**(10.0) - 1.0 ! f(xm)

! Calculando o produto para o novo intervalo
produto = f1*f3

! Atualizando os intervalos em função do novo teste de sinal
if (produto.lt.0.0) then
xs=xm
else
xi=xm
end if
end if

if(opcao.eq.3) then
! Atualizando o contador para verificar quantas vezes uma extremidade se repete

if(xi1.eq.xi) then
contador1 = contador1 + 1
end if

if(xs1.eq.xs) then
contador2 = contador2 + 1
end if
end if

erro = abs(xm1-xm)
iter = iter + 1

end do 

end program falsa_posicao_modificado
